### ISSUE TITLE
Fix S3257 FP: Using discard as lambda argument + within it with _ and __ is not recognized as discard

### ISSUE THREAD

**Issue by ndsvw** [SonarSource/sonar-dotnet#9766]
### Description

Hello,

I have code like this that uses discarding, but a discard in C# has some special behavior, which it may not have in other languages... The type of it can actually matter when using it this way:

This code won't compile at all:
```c#
public class Program
{
	public static void Main()
	{
		Test(_ => {
			_ = DateTime.Now;
		});
	}
	
	public static void Test(Action<int> f){}
}
```
You will get:
> Compilation error (line 8, col 8): Cannot implicitly convert type 'System.DateTime' to 'int'

So, discard kind of gets a type here and you can not "assign values of different types".

Our fallback for such a case is using multiple "_" characters, when using one is not possible. So our code is similar to this:
```c#
public class Program
{
	public static void Main()
	{
		Test(__ => {
			_ = DateTime.Now;
		});
	}
	
	public static void Test(Action<int> f){}
}
```
Unfortunately, SonarDotnet flags this as a problem.
> '__' is not used. Use discard parameter instead.

This is the real code by the way:

<img width="831" height="289" alt="Image" src="https://github.com/user-attachments/assets/f1249a74-9335-4b3c-a6b7-3b1154e3c397" />

Is there a better way to discard in such a case or should "__" just be accepted as discard, too?

### Reproducer

```c#
public class Program
{
	public static void Main()
	{
		Test(__ => {
			_ = DateTime.Now;
		});
	}
	
	public static void Test(Action<int> f){}
}
```

### Product and Version

SonarQube Server - latest or very close to latest (<= 1 month old)

**Comment by pavel-mikula-sonarsource** [comment]
The `Test(_ => {` is not a discard. It's a variable declaration `int _`, where `_` is a variable name. That's why you can not use "discard" syntax inside the method.

Interestingly, `public static void Test(Action<int, int> f) { }` and 
```
        Test((_, _) => {
            _ = DateTime.Now;
            _ = 42;
        });
```
compiles just fine, with actual discard. Roslyn compiler behavior seems to have some backward compatibility tweak to it.

I can't reproduce your scenario to raise the issue there for some reason. What is the language version that your project is using? If you're not sure, add this to any source file and the error message will reveal it
```
#error version
```

<img width="531" height="107" alt="Image" src="https://github.com/user-attachments/assets/0b38207f-6966-4eb3-a1e3-6ab2f1cdf2a7" />

**Comment by ndsvw** [comment]
Hello,

The language version is set to latest via csproj, so it should be 13. It's a .Net Framework 4.8 project by the way, not sure whether that matters.

To be honest: I have not tried out the "reproducer code" with SonarQube. That was just an attempt to simplify it.

This is closer to the real code:
```
protected override void OnStartUp(StartupEventArgs pArgs){
    var host = new HostBuilder().ConfigureAppConfiguration((__, configuration) => {
        _ = configuration.AddXmlFile("test.xml").AddEnvironmentVariables();
    }).Build();
    host.Start();
}
```

**Comment by pavel-mikula-sonarsource** [comment]
I tried this in non-SDK style project targeting .NET Framework 4.8 with `<LangVersion>13</LangVersion>`, but with no luck to raise the issue.
```
public class Repro
{
    protected void OnStartUp()
    {
        var host = new HostBuilder().ConfigureAppConfiguration((__, configuration) => {
            _ = configuration.AddXmlFile("test.xml").AddEnvironmentVariables();
        }).Build();
    }
}

public class HostBuilder
{
    public HostBuilder ConfigureAppConfiguration(System.Action<HostBuilderContext, IConfigurationBuilder> configureDelegate) => null;
    public object Build() => null;
}

public class HostBuilderContext { }

public interface IConfigurationBuilder
{
    IConfigurationBuilder AddXmlFile(string path);
    IConfigurationBuilder AddEnvironmentVariables();
}
```

Can you try to create a small repro project to raise it? To raise the issue, you can use SonarQube for VS extension and activate the rule there, or simply add this NuGet package https://www.nuget.org/packages/SonarAnalyzer.CSharp to your project, and add `.editorconfig` file to activate the rule
```
[*.cs]
dotnet_diagnostic.S3257.severity = warning
```

**Comment by pavel-mikula-sonarsource** [comment]
I'll close this one for now as I couldn't reproduce it. Feel free to reopen it in case you manage to craft a reproducer
